### PR TITLE
Upgrade identifier hashing algorithm and support non phone number identifiers

### LINKED ISSUE #104: Replace current phone hashing with scrypt and add salt in mobile
### Expected Behavior

- [x] Add a salt and switch encryption library
- [x] Prefix the phone number prior to hashing with `tel://`

### Current Behavior

no salt, using sha

### PR DESCRIPTION
**Pull request by rohit-dua**
### Description

The phone hashing algorithm currently used to hash phone numbers is sha3. This is changed to scrypt which is more secure and more customisable. 

The mobile repo is using the `react-native-fast-crypto` while all others are using `scrypt-js`.

The phone numbers are hashed with salted scrypt after prefixing with `tel://` 

### Tested

The packages were tested with local jest test. 
The mobile repo was tested by running app on samsung galaxy phone.

### Related issues

- Fixes #104

### Backwards compatibility

Not backward compatible as the mapping would require changes.

### CODE REVIEW

File: packages/utils/src/attestations.ts [hunk 1]
```diff
@@ -1,40 +1,34 @@
...
-  switch (type) {
-    case IdentifierType.PHONE_NUMBER:
-      return Web3Utils.soliditySha3({ type: 'string', value: identifier })
-    default:
-      return ''
-  }
-}
-
-function attestationMessageToSign(identifier: string, account: string) {
+async function attestationMessageToSign(identifier: string, account: string) {
```
Comment (jmrossy) [inline packages/utils/src/attestations.ts]: Seems problematic that there's a function of the same name in wallet kit which has a different implementation (e.g. this one's not async). @nambrot Was that intentional

File: packages/mobile/src/utils/identity.ts [hunk 2]
```diff
@@ -0,0 +1,36 @@
+import { IdentityUtils as IdentityUtilsJS } from '@celo/utils/'
+import { scrypt as RNscrypt } from 'react-native-fast-crypto'
+
+async function _calculateHashRN(identifier: string) {
```
Comment (jmrossy) [inline packages/mobile/src/utils/identity.ts]: Nit: Remove the underscore, we don't use this naming convention anywhere else

File: packages/mobile/src/utils/identity.ts [hunk 3]
```diff
@@ -0,0 +1,36 @@
+import { IdentityUtils as IdentityUtilsJS } from '@celo/utils/'
```
Comment (jmrossy) [inline packages/mobile/src/utils/identity.ts]: Recommend putting this file in the src/identity folder as utils instead instead of in the utils folder

File: packages/celotool/src/cmds/account.js [hunk 4]
```diff
@@ -0,0 +1,11 @@
+'use strict'
```
Comment (jmrossy) [inline packages/celotool/src/cmds/account.js]: This looks like a compiled TS file, did you mean to include this?

File: packages/utils/src/identity.ts [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+      return IdentifierTypeEnum.PHONE_NUMBER
+  }
+}
+
+function isE164Number(phoneNumber: string) {
+  const E164RegEx = /^\+[1-9][0-9]{1,14}$/
+  return E164RegEx.test(phoneNumber)
+}
+
+async function _calculateHash(identifier: string) {
```
Comment (jmrossy) [inline packages/utils/src/identity.ts]: remote the underscore please

File: packages/utils/src/identity.ts [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+    case 'phoneNumber':
+    case 'phone_number':
+    case 'phone number':
+      return IdentifierTypeEnum.PHONE_NUMBER
+    default:
+      return IdentifierTypeEnum.PHONE_NUMBER
+  }
+}
+
+function isE164Number(phoneNumber: string) {
```
Comment (jmrossy) [inline packages/utils/src/identity.ts]: Please leave this in the phone numbers util file, its more of a phone number concept than an identity one.

File: packages/utils/src/identity.ts (hunk 5 above)
Comment (jmrossy) [inline packages/utils/src/identity.ts]: This function does not need to be async (you do not use await)

File: packages/utils/src/identity.ts [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+    scrypt(
+      Buffer.from(identifier.normalize('NFKC')),
+      Buffer.from(SCRYPT_PARAMS.salt.normalize('NFKC')),
+      SCRYPT_PARAMS.N,
+      SCRYPT_PARAMS.r,
+      SCRYPT_PARAMS.p,
+      SCRYPT_PARAMS.dkLen,
+      (error: any, progress: any, key: any) => {
+        if (error) {
+          throw Error(`Unable to hash ${identifier}, error: ${error}`)
```
Comment (jmrossy) [inline packages/utils/src/identity.ts]: rejecting the promise here would be better than throwing an error

File: packages/contractkit/src/wrappers/Attestations.ts [hunk 8]
```diff
@@ -197,34 +191,34 @@ export class AttestationsWrapper extends BaseWrapper<Attestations> {
...
-   * @param phoneNumber The phone number of the attestation
+   * @param identifier The identifier of the attestation
    * @param account The account of the attestation
    * @param issuer The issuer of the attestation
    * @param code The code received by the validator
    */
-  complete(phoneNumber: string, account: Address, issuer: Address, code: string) {
-    const phoneHash = PhoneNumberUtils.getPhoneHash(phoneNumber)
-    const expectedSourceMessage = attestationMessageToSign(phoneHash, account)
+  async complete(identifier: string, account: Address, issuer: Address, code: string) {
```
Comment (jmrossy) [inline packages/contractkit/src/wrappers/Attestations.ts:206]: Just thinking out loud but maybe it would be better if this and other methods took the identifierHash instead so they wouldn't all need to be converted to async as we recalculate the hash. @nambrot @rohit-dua WDYT?

File: packages/celotool/src/cmds/account/verify.ts [hunk 9]
```diff
@@ -152,7 +152,7 @@ async function verifyCode(
     return
   }
 
-  const tx = await attestations.complete(phoneNumber, account, matchingIssuer, code).send()
+  const tx = await (await attestations.complete(phoneNumber, account, matchingIssuer, code)).send()
```
Comment (jmrossy) [inline packages/celotool/src/cmds/account/verify.ts]: Please break apart to avoid the double await on 1 line here

File: packages/faucet/src/utils.ts [hunk 10]
```diff
@@ -40,18 +40,3 @@ export function generateInviteCode(): {
   const inviteCode = Buffer.from(temporaryPrivateKey.substring(2), 'hex').toString('base64')
   return { address, inviteCode }
 }
-
-// TODO: Copied from @celo/utils, should be removed once usable as a dependency
```
Comment (jmrossy) [inline packages/faucet/src/utils.ts:44]: 👍

File: packages/mobile/jest_setup.ts [hunk 11]
```diff
@@ -12,6 +14,10 @@ const customGlobal: GlobalWithFetchMock = global as GlobalWithFetchMock
 customGlobal.fetch = require('jest-fetch-mock')
 customGlobal.fetchMock = customGlobal.fetch
 
+jest.mock('src/utils/identity', () => ({
```
Comment (jmrossy) [inline packages/mobile/jest_setup.ts]: Why have you mocked this?

File: packages/mobile/src/identity/contactMapping.ts [hunk 12]
```diff
@@ -152,7 +152,10 @@ function* lookupNewRecipients(
 
 async function getAddresses(e164Numbers: string[], attestationsContract: AttestationsType) {
   Logger.debug(TAG, `Get addresses for ${e164Numbers.length} phone numbers`)
-  const phoneHashes = e164Numbers.map((phoneNumber) => getPhoneHash(phoneNumber))
+  const phoneHashPromises = e164Numbers.map((phoneNumber) =>
+    IdentityUtils.identityHash(phoneNumber)
```
Comment (jmrossy) [inline packages/mobile/src/identity/contactMapping.ts]: Does the new crypto lib have a way to batch hashes? Because if not, my intuition is that the promisification of all these tiny hash jobs will actually make things slower. Async management adds a non-trivial amount of overhead. Have you tested the perf? Moreover, each call is going across the RN debugging bridge into native land.

File: packages/utils/package.json [hunk 13]
```diff
@@ -14,22 +14,26 @@
     "test:verbose": "jest --verbose",
     "lint": "tslint -c tslint.json --project ."
   },
-  "files": ["lib/**/*"],
+  "files": [
+    "lib/**/*"
+  ],
   "dependencies": {
+    "@types/scrypt-js": "^2.0.3",
```
Comment (jmrossy) [inline packages/utils/package.json]: @types should be in the dev dependencies section

File: packages/utils/src/identity.ts [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+  salt: 'cel0',
+  N: 512,
+  r: 16,
+  p: 1,
+  dkLen: 32,
+}
+
+function parseIdentifierType(rawType: string): IdentifierTypeEnum {
+  rawType = rawType.toLowerCase()
+  switch (rawType) {
```
Comment (jmrossy) [inline packages/utils/src/identity.ts]: why is this necessary?

**Review by jmrossy — COMMENTED** [review]
@rohit-dua Pretty good first pass, see comments below.

File: packages/utils/src/attestations.ts (hunk 1 above)
Comment (rohit-dua) [reply packages/utils/src/attestations.ts]: I guess the wallet kit function takes identifier hash as parameter while this take the identifier.

File: packages/mobile/jest_setup.ts (hunk 11 above)
Comment (rohit-dua) [reply packages/mobile/jest_setup.ts]: The mobile scrypt library uses native react-native-fast-crypto, so to run the tests locally I mocked it to the js implimentation of scrypt.

File: packages/mobile/src/identity/identity.ts [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+    IdentityUtilsJS.SCRYPT_PARAMS.r,
+    IdentityUtilsJS.SCRYPT_PARAMS.p,
+    IdentityUtilsJS.SCRYPT_PARAMS.dkLen
+  )
+  let hexHash = ''
+  for (const item of result) {
+    hexHash += item.toString(16)
+  }
+  hexHash = hexHash.padStart(64, '0')
+  if (hexHash.startsWith('0x')) {
```
Comment (jmrossy) [inline packages/mobile/src/identity/identity.ts:22]: Consider using https://github.com/celo-org/celo-monorepo/blob/master/packages/utils/src/signatureUtils.ts#L97

File: packages/mobile/src/identity/identity.ts [hunk 16]
```diff
@@ -0,0 +1,84 @@
...
+    const item = result[index]
+    let itemArray = item
+    if (typeof item === 'object') {
+      itemArray = Object.values(item)
+    }
+    for (const hash of itemArray) {
+      hexHash += hash.toString(16)
+    }
+    hexHash = hexHash.padStart(64, '0')
+    hashList[index] = hexHash.startsWith('0x') ? hexHash : '0x' + hexHash
```
Comment (jmrossy) [inline packages/mobile/src/identity/identity.ts:52]: ditto

File: packages/utils/src/identity.ts [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+        if (error) {
+          reject(`Unable to hash ${identifier}, error: ${error}`)
+        } else if (key) {
+          let hexHash = ''
+          for (const item of key) {
+            hexHash += item.toString(16)
+          }
+          // @ts-ignore
+          hexHash = hexHash.padStart(64, '0')
+          if (hexHash.startsWith('0x')) {
```
Comment (jmrossy) [inline packages/utils/src/identity.ts:39]: ditto

**Review by jmrossy — COMMENTED** [review]
Can you post a link to a doc with the performance testing results you gathered?

File: packages/mobile/src/identity/identity.ts [hunk 18]
```diff
@@ -0,0 +1,84 @@
+import { IdentityUtils as IdentityUtilsJS } from '@celo/utils/'
```
Comment (jmrossy) [inline packages/mobile/src/identity/identity.ts:1]: rename this file to utils.ts

File: packages/utils/src/identity.ts [hunk 19]
```diff
@@ -0,0 +1,96 @@
...
+      Buffer.from(SCRYPT_PARAMS.salt.normalize('NFKC')),
+      SCRYPT_PARAMS.N,
+      SCRYPT_PARAMS.r,
+      SCRYPT_PARAMS.p,
+      SCRYPT_PARAMS.dkLen,
+      (error: any, progress: any, key: any) => {
+        if (error) {
+          reject(`Unable to hash ${identifier}, error: ${error}`)
+        } else if (key) {
+          let hexHash = ''
```
Comment (jmrossy) [inline packages/utils/src/identity.ts:33]: what does the logic in this if block do?

File: packages/mobile/src/identity/identity.ts [hunk 20]
```diff
@@ -0,0 +1,84 @@
...
+  type: IdentifierTypeEnum = IdentifierTypeEnum.PHONE_NUMBER
+): Promise<string[]> {
+  if (identifierList.length == 0) {
+    throw Error('Attempting to hash empty identifier list')
+  }
+  const modifiedIdentifierList = IdentityUtilsJS.preprocessPrefixBulk(identifierList, type)
+  return calculateHashRNBulk(modifiedIdentifierList)
+}
+
+export const IdentityUtils = {
```
Comment (jmrossy) [inline packages/mobile/src/identity/identity.ts:80]: Don't wrap these in an object, just export the functions directly

File: packages/mobile/src/identity/identity.ts [hunk 21]
```diff
@@ -0,0 +1,84 @@
+import { IdentityUtils as IdentityUtilsJS } from '@celo/utils/'
+import { scrypt as RNscrypt, scryptBulk as RNscryptBulk } from 'react-native-fast-crypto'
+
+enum IdentifierTypeEnum {
```
Comment (jmrossy) [inline packages/mobile/src/identity/identity.ts:4]: Should this be in @celo/utils instead?
